### PR TITLE
Improve server-related types exported by grpc-js

### PR DESCRIPTION
**Pull request by murgatroid99**
This mainly makes the server call types and `Server#registerService` function signature more specific, the server error types more permissive, and exports some more types needed for writing service implementations.

Also, this fixes a typo in the native library's `index.d.ts` that got brought up in #965. That class is essentially internal, so code written using the typo works. I have added an alias to avoid breaking that.

### CODE REVIEW

File: packages/grpc-js/src/server-call.ts [hunk 1]
```diff
@@ -18,13 +18,14 @@
 import { EventEmitter } from 'events';
 import * as http2 from 'http2';
 import { Duplex, Readable, Writable } from 'stream';
+import * as util from 'util';
```
Comment (cjihrig) [inline packages/grpc-js/src/server-call.ts]: `util.isNumber()` and `util.isString()` are deprecated in favor of `typeof value === 'number'` and `typeof value === 'string'`, respectively.

File: packages/grpc-js/src/server-call.ts [hunk 2]
```diff
@@ -490,21 +495,21 @@ export class Http2ServerCallStream<
...
         ? error.message
         : 'Unknown Error',
-      metadata: error.hasOwnProperty('metadata')
+      metadata: ('metadata' in error && error.metadata !== undefined)
         ? error.metadata
         : new Metadata(),
     };
 
-    if (error.hasOwnProperty('code') && Number.isInteger(error.code)) {
+    if ('code' in error && util.isNumber(error.code) && Number.isInteger(error.code)) {
```
Comment (cjihrig) [inline packages/grpc-js/src/server-call.ts]: The `util.isNumber()` (or `typeof error.code === 'number')` check shouldn't be necessary due to the `Number.isInteger()` call.

Comment (murgatroid99) [reply packages/grpc-js/src/server-call.ts]: I think Typescript was not happy with calling `Number.isInteger` on that value without verifying that it was a number.

Comment (cjihrig) [reply packages/grpc-js/src/server-call.ts]: I think you're right: https://github.com/Microsoft/TypeScript/issues/21199 😕

Comment (cjihrig) [reply packages/grpc-js/src/server-call.ts]: I think you're right: https://github.com/Microsoft/TypeScript/issues/21199 😕

**Pull request merged**